core/src/mock_alpenglow_consensus.rs

```diff
@@ -0,0 +1,593 @@
+            }
+        }
+
+        let mut stake_weighted_delay = [0f64; 4];
+        let mut percent_collected = [0f64; 4];
+
+        for i in 0..4 {
+            if total_voted_stake[i] > 0 {
+                stake_weighted_delay[i] = total_delay_ms[i] as f64 / total_voted_stake[i] as f64;
+            }
+            percent_collected[i] = 100.0 * total_voted_stake[i] as f64 / total_staked as f64;
+
+            info!(
+                "{:?}: got {} % of total stake collected, stake-weighted delay is {}ms",
+                VotorMessageType::try_from(i as u64).unwrap(), // this unwrap is ok since i is in static range
+                percent_collected[i],
+                stake_weighted_delay[i]
+            );
+        }
+        datapoint_info!(
+            "mock_alpenglow",
+            ("total_peers", peers.len(), f64),
+            ("packets_collected_notarize", total_voted_nodes[0], f64),
+            (
+                "percent_stake_collected_notarize",
+                percent_collected[0],
+                f64
+            ),
+            ("weighted_delay_ms_notarize", stake_weighted_delay[0], f64),
+            ("packets_collected_notarize_cert", total_voted_nodes[1], f64),
+            (
+                "percent_stake_collected_notarize_cert",
+                percent_collected[1],
+                f64
+            ),
+            (
+                "weighted_delay_ms_notarize_cert",
+                stake_weighted_delay[1],
+                f64
+            ),
+            ("packets_collected_finalize", total_voted_nodes[2], f64),
+            (
+                "percent_stake_collected_finalize",
+                percent_collected[2],
+                f64
+            ),
+            ("weighted_delay_ms_finalize", stake_weighted_delay[2], f64),
+            ("packets_collected_finalize_cert", total_voted_nodes[3], f64),
+            (
+                "percent_stake_collected_finalize_cert",
+                percent_collected[3],
+                f64
+            ),
+            (
+                "weighted_delay_ms_finalize_cert",
+                stake_weighted_delay[3],
+                f64
+            ),
+        );
+    }
+
+    pub(crate) fn signal_new_slot(&mut self, slot: Slot, interval: Slot) {
+        // constraint interval to be above 3 to keep current logic working
+        let interval = interval.max(3);
+        let phase = slot % interval;
+        // this is currently set up to work over a course of 3 slots.
+        // it will only work correctly if we do not actually fork within those 3 slots, and
+        // are consistently casting votes for all 3.
+        match phase {
+            // Clear the stats for the previous slots in preparation for the new one
+            0 => {
+                // prepare to receive votes for the next slot in advance
+                // in case we are really late getting shreds
+                self.prepare_to_receive(slot + 1);
```

Comment:
We might need to test sending all-to-all packets on every slot, so why do we need a whole slot to prepare_to_receive? Can we remember on which slot we reported last time, if the slot changed, report and clear stats, then count the packets for the new slot?

Reply:
Yes, this can be pipelined, but it will make logic somewhat more complex. I can implement that too if needed.